lib/Sema/CSSimplify.cpp

```diff
@@ -3235,6 +3235,12 @@ ConstraintSystem::matchFunctionTypes(FunctionType *func1, FunctionType *func2,
       return getTypeMatchFailure(locator);
   }
 
+  // () -> sending T can be a subtype of () -> T... but not vis-a-versa.
+  if (func1->hasSendingResult() != func2->hasSendingResult() &&
+      (!func1->hasSendingResult() || kind < ConstraintKind::Subtype)) {
+    return getTypeMatchFailure(locator);
```

Comment:
I think both of these new checks require a tailored fix to explain what exactly went wrong because otherwise users would be left to figure out why two function types didn't match which might not always be obvious for complex types... Doesn't have to be implemented in this PR though.